### PR TITLE
fix(iterable): refactored destination code

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

- This pull request (PR) encompasses a complete refactoring of the Iterable Destination code, incorporating numerous enhancements and optimizations. The aim of this refactor is to enhance the overall performance, maintainability, and extensibility of the codebase

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/iterable/transform.js [hunk 1]
```diff
@@ -228,47 +233,89 @@ function batchEvents(arrayChunks) {
...
+      };
+      batchedResponseList.push(
+        getSuccessRespEvents(
+          batchEventResponse.batchedRequest,
+          batchEventResponse.metadata,
+          batchEventResponse.destination,
+          true,
+        ),
+      );
+    }
```
Comment (ItsSudip) [inline src/v0/destinations/iterable/transform.js:318]: This part is repeating can't we move this to a function?

File: src/v0/destinations/iterable/transform.js [hunk 2]
```diff
@@ -191,6 +193,11 @@ function process(event) {
 function batchEvents(arrayChunks) {
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/v0/destinations/iterable/transform.js:226]: it was written around an year ago, so i haven't changed anything. probably with the next part of refactoring we can convert it to ES6

**Review by ItsSudip — COMMENTED** [review]
Can you add some test cases covering the scenario?

**Review by anantjain45823 — DISMISSED** [review]

File: src/v0/destinations/iterable/config.js [hunk 3]
```diff
@@ -64,15 +64,20 @@ const ConfigCategory = {
     endpoint: `${BASE_URL}catalogs`,
   },
 };
+
 const mappingConfig = getMappingConfig(ConfigCategory, __dirname);
+
 const IDENTIFY_MAX_BATCH_SIZE = 1000;
-const TRACK_MAX_BATCH_SIZE = 8000;
+const IDENTIFY_MAX_BODY_SIZE = 4000000;
```
Comment (utsabc) [inline src/v0/destinations/iterable/config.js:71]:
```suggestion
const IDENTIFY_MAX_BODY_SIZE_IN_BYTES = 4000000;
```

Comment (mihir-4116) [reply src/v0/destinations/iterable/config.js]: updated

**Review by utsabc — CHANGES_REQUESTED** [review]

File: src/v0/destinations/iterable/transform.js [hunk 4]
```diff
@@ -181,162 +178,92 @@ function processSingleMessage(message, destination) {
...
         if (event.message.statusCode) {
           // already transformed event
-          getEventChunks(event, identifyEventChunks, trackEventChunks, eventResponseList);
+          filterEvents(
+            event,
+            catalogEvents,
+            updateUserEvents,
+            registerDeviceOrBrowserEvents,
+            trackEvents,
+            eventResponseList,
```
[the remark this answers is not in the document]
Comment (mihir-4116) [reply src/v0/destinations/iterable/transform.js]: Updated

File: src/v0/destinations/iterable/transform.js [hunk 5]
```diff
@@ -2,375 +2,197 @@ const _ = require('lodash');
...
     default:
-      throw new InstrumentationError(`Message type ${messageType} not supported`);
+      throw new InstrumentationError(`Message type ${eventType} not supported`);
   }
-  const response = responseBuilderSimple(message, category, destination);
+};
+
+const process = (event) => {
+  const { message, destination } = event;
+  const messageType = message.type.toLowerCase();
```
Comment (utsabc) [inline src/v0/destinations/iterable/transform.js:148]:
```suggestion
  const messageType = message.type?.toLowerCase();
```

Comment (mihir-4116) [reply src/v0/destinations/iterable/transform.js]: updated

File: src/v0/destinations/iterable/util.js [hunk 6]
```diff
@@ -1,230 +1,692 @@
...
+ * Common function to build screen or page event payload
+ * @param {*} message
+ * @param {*} destination
+ * @param {*} category
+ * @returns
+ */
+const pageOrScreenEventPayloadBuilder = (message, destination, category) => {
   let rawPayload = {};
-  if (destination.Config.trackAllPages) {
+  const eventType = message.type.toLowerCase();
```
Comment (utsabc) [inline src/v0/destinations/iterable/util.js:142]: I think we should add a common validation for message type in `process` func

File: src/v0/destinations/iterable/transform.js [hunk 7]
```diff
@@ -2,375 +2,197 @@ const _ = require('lodash');
...
+        let responses = await process(event);
+        responses = Array.isArray(responses) ? responses : [responses];
+        responses.forEach((response) => {
+          transformedPayloads.push({
+            message: response,
+            metadata: event.metadata,
+            destination: event.destination,
+          });
+        });
+        return transformedPayloads;
```
Comment (utsabc) [inline src/v0/destinations/iterable/transform.js:187]: This logic will add same metadata twice if `responses` is array

File: src/v0/destinations/iterable/util.js [hunk 8]
```diff
@@ -1,230 +1,692 @@
...
+  const updateUserEvents = categorizeEventList
+    .filter((event) => event.type === 'updateUser')
+    .map((event) => event.data);
+  const registerDeviceOrBrowserTokenEventsArray = categorizeEventList.filter(
+    (event) => event.type === 'registerDeviceOrBrowser',
+  );
+  const registerDeviceOrBrowserTokenEvents = mapRegisterDeviceOrBrowserTokenEventsWithJobId(
+    registerDeviceOrBrowserTokenEventsArray,
+  );
+
```
Comment (utsabc) [inline src/v0/destinations/iterable/util.js:639]: Need some clarity here what are we trying to do

**Comment by mihir-4116** [comment]
closing this pr as it's showing more changes then expected due to commit mess up
created new pr [here](https://github.com/rudderlabs/rudder-transformer/pull/2390)

**Pull request closed without merging**